### ISSUE TITLE
What is the compression profile of the proposed handling of revocation 2 merkle tree leaves

### ISSUE THREAD

**Issue by swcurran** [hyperledger/indy-node#1596]
See #1595 for details of the plan for revocation 2.

We need a technical spike done to verify our assumptions about the proposed approach to storing the leaves of the merkle tree on the ledger.  Notable, the leaves, a list of 64-bit unsigned ints with each bit representing 1 credential's revocation status.  The request is to do something like:

- define a way to store the leaves of the merkle tree based on a parameter of the number of layers (1-7).
- For a test, define a percent of the number of credentials that will be revoked, ranging from say 10 to 50%.
- Randomly generate the revocation status of all the credentials (the bit) based on the percent.
- gzip the result.
- output the parameters of the test and the size of the resulting compressed chunk of data.

The code is purely a tech spike and need not be production worthy.  It should probably be pushed to the repo for review by the maintainers and for possible use in the future - for example, playing with different ways of storing the (uncompressed) set of leaves, and testing different compression algorithms.

**Comment by dhh1128** [comment]
Can I claim this ticket, or was someone else planning to?

**Comment by swcurran** [comment]
Sold.  You've got it, Daniel.  Thanks!

**Comment by dhh1128** [comment]
I've investigated compression in conjunction with some other benchmarking; see #1597

**Comment by swcurran** [comment]
@dhh1128 - did you do any investigation of the compression of just the leaves (bits) without the rest of the merkle tree?  I think the compression included the entire tree, but I might be wrong.  Please clarify.

**Comment by dhh1128** [comment]
No, I did not study compression of just the leaves; the numbers I published were for the tree as a whole. Compression of pure leaves will be *very* high -- over 90%, likely, or even more -- until the revocations become dense. And since we're talking about compressing data that is already slimmed down to one bit per revocation, it will produce wonderfully small files. So a quick estimate: for a revocation herd of 2M, we'd need 2M bits, uncompressed. That's 256 KB. Then if you compress 90%, you're talking about maybe a 25KB file.

I'll come back and post concrete numbers for different scenarios, but that's the sort of results I expect.
